File: src/main/java/com/jore/epoc/bo/orders/BuildStorageOrder.java
package com.jore.epoc.bo.orders;

import org.hibernate.annotations.CompositeType;

import com.jore.datatypes.money.Money;
import com.jore.epoc.bo.Storage;
import com.jore.epoc.bo.accounting.FinancialAccounting;
import com.jore.epoc.bo.message.MessageLevel;

import jakarta.persistence.AttributeOverride;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;

@Entity
public class BuildStorageOrder extends AbstractSimulationOrder {
    private Integer capacity;
    private Integer timeToBuild;
    @AttributeOverride(name = "amount", column = @Column(name = "fixed_cost_amount"))
    @AttributeOverride(name = "currency", column = @Column(name = "fixed_cost_currency"))
    @CompositeType(com.jore.datatypes.hibernate.MoneyCompositeUserType.class)
    private Money constructionCosts;
    @AttributeOverride(name = "amount", column = @Column(name = "variable_cost_amount"))
    @AttributeOverride(name = "currency", column = @Column(name = "variable_cost_currency"))
    @CompositeType(com.jore.datatypes.hibernate.MoneyCompositeUserType.class)
    private Money constructionCostsPerUnit;
    @AttributeOverride(name = "amount", column = @Column(name = "inventory_cost_amount"))
    @AttributeOverride(name = "currency", column = @Column(name = "inventory_cost_currency"))
    @CompositeType(com.jore.datatypes.hibernate.MoneyCompositeUserType.class)
    private Money inventoryManagementCost;

    @Override
    public void execute() {
        if (getCompany().getAccounting().checkFunds(getAmount(), getExecutionMonth().atEndOfMonth())) {
            addStorage();
            book(getExecutionMonth().atDay(FIRST_OF_MONTH), "Built storage", FinancialAccounting.REAL_ESTATE, FinancialAccounting.BANK, constructionCosts.add(constructionCostsPerUnit.multiply(capacity)));
            addMessage(MessageLevel.INFORMATION, "StorageBuilt", capacity, getExecutionMonth());
            setExecuted(true);
        } else {
            addMessage(MessageLevel.WARNING, "NoStorageDueToFunds", getExecutionMonth(), getAmount(), getCompany().getAccounting().getBankBalance(getExecutionMonth().atEndOfMonth()));
        }
    }

    @Override
    public Money getAmount() {
        return constructionCosts.add(constructionCostsPerUnit.multiply(capacity));
    }

    public Money getInventoryManagementCost() {
        return inventoryManagementCost;
    }

    @Override
    public int getSortOrder() {
        return 2;
    }

    @Override
    public String getType() {
        return "Build Storage";
    }

    public void setCapacity(Integer capacity) {
        this.capacity = capacity;
    }

    public void setConstructionCost(Money constructionCosts) {
        this.constructionCosts = constructionCosts;
    }

    public void setConstructionCostPerUnit(Money constructionCostsPerUnit) {
        this.constructionCostsPerUnit = constructionCostsPerUnit;
    }

    public void setInventoryManagementCost(Money inventoryManagementCost) {
        this.inventoryManagementCost = inventoryManagementCost;
    }

    public void setTimeToBuild(Integer timeToBuild) {
        this.timeToBuild = timeToBuild;
    }

    private void addStorage() {
        Storage storage = new Storage();
        storage.setCapacity(capacity);
        storage.setStorageStartMonth(getExecutionMonth().plusMonths(timeToBuild));
        storage.setInventoryManagementCost(inventoryManagementCost);
        getCompany().addStorage(storage);
    }
}

File: EpocServer/Sources/App/Models/Orders/BuildStorageOrder.swift
import Foundation

public final class BuildStorageOrder: SimulationOrder {
    public var id: Int64?
    public var executionMonth: YearMonth = .of(2020, 1)
    public var isExecuted: Bool = false
    public weak var company: Company!
    public var capacity: Int = 0
    public var timeToBuild: Int = 0
    public var constructionCost: Money = Money.of("CHF", 1000000)
    public var constructionCostPerUnit: Money = Money.of("CHF", 100)
    public var inventoryManagementCost: Money = Money.of("CHF", 500000)

    public var sortOrder: Int { 2 }
    public var type: String { "Build Storage" }

    public func getAmount() -> Money {
        constructionCost.add(constructionCostPerUnit.multiply(capacity))
    }

    public func execute() {
        if company.accounting.checkFunds(getAmount(), valueDate: executionMonth.atEndOfMonth()) {
            addStorage()
            book(bookingDate: executionMonth.atDay(1), bookingText: "Built storage",
                 debitAccount: FinancialAccounting.REAL_ESTATE, creditAccount: FinancialAccounting.BANK,
                 amount: constructionCost.add(constructionCostPerUnit.multiply(capacity)))
            addMessage(.information, "StorageBuilt", capacity, executionMonth)
            isExecuted = true
        } else {
            addMessage(.warning, "NoStorageDueToFunds", executionMonth, getAmount(), company.accounting.getBankBalance(executionMonth.atEndOfMonth()))
        }
    }

    private func addStorage() {
        let storage = Storage()
        storage.capacity = capacity
        storage.storageStartMonth = executionMonth.plusMonths(timeToBuild)
        storage.inventoryManagementCost = inventoryManagementCost
        company.addStorage(storage)
    }
}
